### PR TITLE
IDA Convfit get lorentzian width from ipf

### LINKED ISSUE #29254: ConvFit get Lorentzian width from IPF
**Original reporter:** ISIS/Spencer

Lorentzian does not get resolution width from IPF for Guess parameter. This was in the old version but not transferred
### Expected behavior
Lorentzian width should be taken from the IPF.

### PR DESCRIPTION
**Pull request by Harrietbrown**
**Description of work.**

This PR makes it so that the Convfit tab in the indirct data analysis gui can get it's initial guess paramiters for it's lorenzian from the resolution IPF when it is loaded. 
Also as part of finding a solution to this I have also added a propper overload for IFunction getError and SetError by paramiter name instead of just index.

**To test:**

1. get data from #24664 
1. Open workbench
2. Open Indirect->Data Analysis
3. Go to Convfit
4. In the function browser add a lorenzian
5. Load a sample resolution pair
6. The function browser shoud have it's FWHM changed to 0.0175, the rage markers on the plor should have been updated to reflect this.
7. change to multiple input and do the same with indevidual spectra and combinations of different workspaces.

Fixes #29254 

The changes to convfit has release notes, but the overload for getError and setError are internal only and do not require them.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by robertapplin** [comment]
This has merge conflicts @Harrietbrown

**Comment by robertapplin** [comment]
I am getting a `Vector subscript out of range` error when I load in this data and select 1 Lorentzian:
[irs26173_graphite002_res.zip](https://github.com/mantidproject/mantid/files/5531221/irs26173_graphite002_res.zip)

**Comment by Harrietbrown** [comment]
I'm not able to recreate a `Vector subscript out of range` error on my end, could you give some more specific instructions

File: qt/widgets/common/src/FunctionModel.cpp [hunk 1]
```diff
@@ -505,7 +505,7 @@ void FunctionModel::setResolutionFromWorkspace(IFunction_sptr fun) {
         auto analyser = inst->getStringParameter("analyser");
         if (analyser.size() > 0) {
           auto comp = inst->getComponentByName(analyser[0]);
-          if (comp) {
+          if (comp->hasParameter("resolution")) {
```
Comment (robertapplin) [inline qt/widgets/common/src/FunctionModel.cpp:508]: Might we need to check that `comp` exists before we check `hasParameter`?

**Review by robertapplin — CHANGES_REQUESTED** [review]

File: qt/widgets/common/src/FunctionModel.cpp (hunk 1 above)
Comment (robertapplin) [reply qt/widgets/common/src/FunctionModel.cpp:508]: Nevermind, it looks like this will work fine.

**Review by robertapplin — DISMISSED** [review]

**Review by robertapplin — DISMISSED** [review]
Looks good 👍

File: qt/widgets/common/src/FunctionModel.cpp [hunk 2]
```diff
@@ -487,6 +489,37 @@ void FunctionModel::updateGlobals() {
...
+  auto n = fun->getNumberDomains();
+  if (n > 1) {
+    for (size_t index = 0; index < n; index++) {
+      setResolutionFromWorkspace(fun->getFunction(index));
+    }
+  } else {
+    if (fun->hasAttribute("f0.Workspace")) {
+      std::string wsName = fun->getAttribute("f0.Workspace").asString();
+      if (AnalysisDataService::Instance().doesExist(wsName)) {
+        const auto ws =
```
Comment (martyngigg) [inline qt/widgets/common/src/FunctionModel.cpp:502]: I'm finding this a little difficult to parse due to the heavy nesting. Could we maybe extract out another method
```cpp
setResolutionFromWorkspace(const MatrixWorkspace & workspace)
```

that could be called as

```cpp
setResolutionFromWorkspace(AnalysisDataService::Instance().retrieveWS<MatrixWorkspace>(wsName));
```
?

File: qt/widgets/common/src/FunctionModel.cpp [hunk 3]
```diff
@@ -487,6 +489,37 @@ void FunctionModel::updateGlobals() {
...
+            AnalysisDataService::Instance().retrieveWS<MatrixWorkspace>(wsName);
+        auto inst = ws->getInstrument();
+        auto analyser = inst->getStringParameter("analyser");
+        if (analyser.size() > 0) {
```
Comment (martyngigg) [inline qt/widgets/common/src/FunctionModel.cpp:506]: With the new extracted method we could do an early return with
```cpp
if(analyser.empty())
  return
```
to further reduce the indentation and improve readability?

**Review by martyngigg — CHANGES_REQUESTED** [review]
A nice addition to the function layer to mirror the access by name for the paramters.

Are there any further tests that could be added around CompositeFunction, ParamFunction to test the new behaviour?

**Review by martyngigg — APPROVED** [review]
Thanks for adding those tests.

**Comment by martyngigg** [comment]
Unrelated test failure.

**Pull request merged**